### PR TITLE
feat(elasticloadbalancing): rename 'sslCertificateId' property of LB listener to 'sslCertificateArn'; deprecate sslCertificateId  property

### LINKED ISSUE #9303: [elasticloadbalancing] Clarify expected value of sslCertificateId property of LoadBalancerListener is an ARN
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-elasticloadbalancing.LoadBalancerListener.html#sslcertificateid

I was attempting to replicate some existing AWS resources in a CDK app and one of the resources was a Classic Load Balancer which had an HTTPS listener and was using an ACM certificate. The documentation describes the `sslCertificateId` property of `LoadBalancerListener` as the "ID of the SSL Certificate", and the data type of the field is a `string`, so I supplied the ACM identifier. When I tried `cdk deploy`, I eventually got an HTTP 400 error "Server Certificate not found for the key". I verified that I had no typos and was left a little confused as to why the `sslCertificateId` was not found, so I went to the CLI documentation looking for additional information on the API call being used and any restrictions on the input. I landed on the AWS CLI documentation for [set-load-balancer-listener-ssl-certificate](https://docs.aws.amazon.com/cli/latest/reference/elb/set-load-balancer-listener-ssl-certificate.html), where the definition of `--ssl-certificate-id` states that this is the ARN of the SSL Certificate. 

I would like the CDK documentation updated to reflect this, and explicitly state that `sslCertificateId` refers to the ARN of the certificate and not just the identifier, as simply referring to it as an ID or identifier is a bit ambiguous for those not already aware that the expected value is an ARN.

---

This is a 📕 documentation issue

### PR DESCRIPTION
**Pull request by OksanaH**
The property `sslCertificateId` of the LoadBalancer listener actually means sslCertificateArn. So as suggested in #9303, I have deprecated `sslCertificateId` and replaced it by `sslCertificateArn` to better reflect its actual meaning.

fixes #9303

### CODE REVIEW

File: packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts [hunk 1]
```diff
@@ -187,10 +187,18 @@ export interface LoadBalancerListener {
   readonly policyNames?: string[];
 
   /**
-   * ID of SSL certificate
+   * the ARN of SSL certificate
```
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts:190]:
```suggestion
   * the ARN of the SSL certificate
```

File: packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts [hunk 2]
```diff
@@ -187,10 +187,18 @@ export interface LoadBalancerListener {
...
+   * @deprecated - use sslCertificateArn instead
    */
   readonly sslCertificateId?: string;
 
+  /**
+   * the ARN of the SSL certificate
+   *
+   * @default none
```
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts:198]:
```suggestion
   * @default - none
```

File: packages/@aws-cdk/aws-elasticloadbalancing/README.md [hunk 3]
```diff
@@ -44,3 +44,6 @@ lb.addListener({
     allowConnectionsFrom: [mySecurityGroup]
 });
 ```
+
+The `sslCertificateId` property of the `LoadBalancerListener` class has been deprecated.
+It is recommended to use `sslCertificateArn` property instead.
```
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancing/README.md:49]: I think we can skip the README update in this case, as it describes a change more so than the current state. Can you revert these changes?

File: packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts [hunk 4]
```diff
@@ -266,6 +274,7 @@ export class LoadBalancer extends Resource implements IConnectable {
   public addListener(listener: LoadBalancerListener): ListenerPort {
     const protocol = ifUndefinedLazy(listener.externalProtocol, () => wellKnownProtocol(listener.externalPort));
     const instancePort = listener.internalPort || listener.externalPort;
+    const sslCertificateArn = listener.sslCertificateId || listener.sslCertificateArn;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts:277]: I'd prefer we default to using the newer property over the old one.

Bonus points for creating some error handling in the case where a user specifies both.

```suggestion
    const sslCertificateArn = listener.sslCertificateArn || listener.sslCertificateId;
```

**Review by njlynch — CHANGES_REQUESTED** [review]
Thanks for the contribution! I made a few minor suggestions for improvements.

File: packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts (hunk 4 above)
Comment (OksanaH) [reply packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts]: @njlynch Thanks for the comments. When I was writing a unit test to test error handling, there was an error "Property 'toThrowError' does not exist on type 'StackInspector'". I also tried to add @aws-cdk/assert/jest, rebuilt etc, but the error persisted. Any idea why this is? Thanks.

File: packages/@aws-cdk/aws-elasticloadbalancing/README.md (hunk 3 above)
Comment (OksanaH) [reply packages/@aws-cdk/aws-elasticloadbalancing/README.md]: Do I need to do anything else here, because a check is failing due to there not being changes to the README.md?

File: packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts (hunk 4 above)
Comment (njlynch) [reply packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts]: Depends. What did the test look like? Typically, it would look something like this:

```ts
expect(() => lb.addListener(...)).toThrow(/error message here/);
```

Comment (OksanaH) [reply packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts]: So, in the load-balancer.ts I had
<pre><code>if (listener.sslCertificateArn && listener.sslCertificateId) {
      throw new Error('"sslCertificateId" is deprecated, please use "sslCertificateArn" only.');
    }</code></pre>

The relevant part of test was like you said:

<pre><code>expect(() =>
    lb.addListener({
      externalPort: 80,
      internalPort: 8080,
      sslCertificateArn: sslCertificateArn,
    })).toThrow(/"sslCertificateId" is deprecated, please use "sslCertificateArn" only./);</code></pre>

Comment (njlynch) [reply packages/@aws-cdk/aws-elasticloadbalancing/lib/load-balancer.ts]: Ah, yes, I see now.

The `expect` is currently coming from `@aws-cdk/assert`, which is overriding the default jest expect. This is a holdover from the older nodeunit-style tests here.

The following should work:
1. Import `@aws-cdk/assert/jest` (as you said you tried).
2. Remove the imports from `'@aws-cdk/assert'`.
3. Rewrite the existing expects from `expect(stack).to(haveResource(....));` to `expect(stack).toHaveResource(...);`.

Can you give that a shot?

**Review by njlynch — APPROVED** [review]
Looks great, thanks!

**Comment by OksanaH** [comment]
@njlynch, looks like the build has failed - is there something that I need to do to resolve this issue? Thanks

**Review by njlynch — APPROVED** [review]

**Comment by njlynch** [comment]
> looks like the build has failed - is there something that I need to do to resolve this issue? Thanks

Our `@aws-cdk/assert` library was renamed to `@aws-cdk/assert-internal`, which caused the build to fail. I fixed it for you in the diff, and it's merging cleanly now. Thanks again for the contribution!
